src/test/java/com/amazon/ion/impl/bin/IonManagedWriter_1_1_Test.kt

```diff
@@ -97,4 +143,612 @@ internal class IonManagedWriter_1_1_Test {
+                    body = {
+                        variable(0)
+                    },
+                    expectedBody = "x"
+                ),
+                case(
+                    "multiple variables",
+                    signature = listOf("x", "y", "z").map(Parameter::exactlyOneTagged),
+                    expectedSignature = "(x y z)",
+                    body = {
+                        list {
+                            variable(0)
+                            variable(1)
+                            variable(2)
+                        }
+                    },
+                    expectedBody = "[x,y,z]"
+                ),
+                case(
+                    "nested expressions in body",
+                    body = {
+                        list {
+                            sexp { int(1) }
+                            struct {
+                                fieldName("foo")
+                                int(2)
+                            }
+                        }
+                    },
+                    expectedBody = "[(make_sexp [1]),{foo:2}]"
+                ),
+
+            )
+        }
+    }
+
+    @MethodSource(TestWritingMacroDefinitions.THE_METHOD)
+    @ParameterizedTest(name = "a macro definition with {0}")
+    fun testWritingMacroDefinitions(description: String, macro: Macro, expectedSignatureAndBody: String) {
+        val expected = """
+            $ion_1_1
+            $ion_encoding::((macro_table (macro null $expectedSignatureAndBody)))
+        """.trimIndent()
+
+        val actual = write { addMacro(macro) }
+
+        assertEquals(expected, actual)
+    }
+
+    @Test
+    fun `writeObject() should write something with a macro representation`() {
+        val expected = """
+            $ion_1_1
+            $ion_encoding::((macro_table (macro com_amazon_ion_impl_bin_IonManagedWriter_1_1_Test_Point2D (x y) {x:x,y:y})))
+            (:com_amazon_ion_impl_bin_IonManagedWriter_1_1_Test_Point2D 2 4)
+        """.trimIndent()
+
+        val actual = write {
+            writeObject(Point2D(2, 4))
+        }
+
+        assertEquals(expected, actual)
+    }
+
+    @Test
+    fun `writeObject() should write something without a macro representation`() {
+        val expected = """
+            $ion_1_1
+            Red
+            Yellow
+            Green
+            Blue
+        """.trimIndent()
+
+        val actual = write {
+            Colors.entries.forEach {
+                    color ->
+                writeObject(color)
+            }
+        }
+
+        assertEquals(expected, actual)
+    }
+
+    private data class Point2D(val x: Long, val y: Long) : WriteAsIon {
+        companion object {
+            // This is a very long macro name, but by using the qualified class name,
+            // there is almost no risk of having a name conflict with another macro.
+            private val MACRO_NAME = Point2D::class.qualifiedName!!.replace(".", "_")
```

Comment:
🗺️ It would be even safer to replace `.` with `$`, but that was not playing nicely with the expected output in the test case because Kotlin uses `$` to indicate string interpolation.